cmd/dex/serve.go

```diff
@@ -73,22 +73,49 @@ func commandServe() *cobra.Command {
 	return cmd
 }
 
-func listenAndShutdownGracefully(logger log.Logger, gr *run.Group, srv *http.Server, name string) error {
-	l, err := net.Listen("tcp", srv.Addr)
+type serverRunner struct {
```

Comment:
TBH I'm not sure I like this structure. It complicates the code IMHO, decreases visibility while it only spares a few lines of code.

Here is how I usually do it: https://github.com/sagikazarmark/modern-go-application/blob/master/cmd/modern-go-application/main.go#L194-L214

It does come with some duplications, but keeps the code more readable in my opinion.

WDYT?

Reply:
1. About the bug: Yes, it is caused by not wrapping the listener with TLS. `ServeTLS` solves the problem.
2. Ok, I will try to simplify it. My major intention was to hide some common logic inside a black-box, e.g., logging, listener creation. However, I have no objection to try other approaches.

Reply:
See #2019